formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/linearization/SsaConverter.kt

```diff
@@ -2,33 +2,39 @@ package org.jetbrains.kotlin.formver.core.linearization
 
 import org.jetbrains.kotlin.KtSourceElement
 import org.jetbrains.kotlin.formver.common.SnaktInternalException
+import org.jetbrains.kotlin.formver.core.names.SsaVariableName
+import org.jetbrains.kotlin.formver.viper.SymbolicName
 import org.jetbrains.kotlin.formver.viper.ast.Declaration
 import org.jetbrains.kotlin.formver.viper.ast.Exp
 
-class SsaConverter(val source: KtSourceElement? = null) {
-    private val assignments: MutableMap<Declaration.LocalVarDecl, Exp> = mutableMapOf()
+class SsaConverter(
+    val source: KtSourceElement? = null,
+    val variableIndex: MutableMap<SymbolicName, Int> = mutableMapOf()
+) {
+    private val assignments: MutableList<Triple<Declaration.LocalVarDecl, Int, Exp>> = mutableListOf()
     private var body: Exp? = null
 
     fun asExp(): Exp {
         if (body == null) throw SnaktInternalException(
             source,
             "Empty body cannot be converted into let chain"
         )
-        return assignments.entries.toList().foldRight(body as Exp) { (decl, expr), acc ->
-            Exp.LetBinding(decl, expr, acc)
+        return assignments.foldRight(body as Exp) { (decl, ssaIdx, expr), acc ->
+            Exp.LetBinding(decl.copy(name = SsaVariableName(decl.name, ssaIdx)), expr, acc)
         }
     }
 
     fun addAssignment(decl: Declaration.LocalVarDecl, varExp: Exp) {
-        // TODO: Allow this case using SSA transformation
-        if (assignments.containsKey(decl)) throw SnaktInternalException(
-            source,
-            "Found duplicate variable declaration while constructing a let-chain. Declaration is $decl"
-        )
-        assignments[decl] = varExp
+        variableIndex[decl.name] = variableIndex[decl.name]?.plus(1) ?: 0
+        assignments.add(Triple(decl, variableIndex[decl.name] as Int, varExp))
     }
 
     fun addBody(body: Exp) {
         this.body = body
     }
+
+    fun resolveVariableName(name: SymbolicName): SymbolicName =
+    // Fall back to parameter if no assignment in map is found
+        // TODO: Fall back to global value numbering before falling back to parameter
+        assignments.lastOrNull { it.first.name == name }?.let { SsaVariableName(name, it.second) } ?: name
```

Comment:
Could you be more specific about why this is the desired behaviour?  And why not use `variableIndex`?